src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -323,3 +500,237 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
   fn_fail:
     goto fn_exit;
 }
+
+int MPIR_Typerep_op(void *source_buf, int source_count, MPI_Datatype source_dtp,
+                    void *target_buf, int target_count, MPI_Datatype target_dtp,
+                    MPI_Op acc_op, int src_kind)
+{
+
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_op_function *uop = NULL;
+    MPI_Aint source_dtp_size = 0, source_dtp_extent = 0, max_pack_bytes, actual_pack_bytes;
+    int is_empty_source = FALSE;
+    int is_supported = 0;
+    struct iovec *typerep_vec = NULL;
+    int i;
+    MPI_Aint vec_len = 0, type_extent = 0, type_size = 0, src_type_stride = 0;
+    MPI_Datatype type;
+    MPIR_Datatype *dtp;
+    MPI_Aint curr_len;
+    void *curr_loc;
+    int accumulated_count;
+    char *src_ptr = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_TYPEREP_RMA_REDUCE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_TYPEREP_RMA_REDUCE);
+
+    /* first Judge if source buffer is empty */
+    if (acc_op == MPI_NO_OP)
+        is_empty_source = TRUE;
+
+    if (is_empty_source == FALSE) {
+        MPIR_Assert(MPIR_DATATYPE_IS_PREDEFINED(source_dtp));
+        MPIR_Datatype_get_size_macro(source_dtp, source_dtp_size);
+        MPIR_Datatype_get_extent_macro(source_dtp, source_dtp_extent);
+    }
+
+    if ((HANDLE_IS_BUILTIN(acc_op))
+        && ((*MPIR_OP_HDL_TO_DTYPE_FN(acc_op)) (source_dtp) == MPI_SUCCESS)) {
+        is_supported = MPIR_Typerep_reduce_is_supported(acc_op, source_dtp);
+    } else {
+        /* --BEGIN ERROR HANDLING-- */
+        mpi_errno = MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE,
+                                         __func__, __LINE__, MPI_ERR_OP,
+                                         "**opnotpredefined", "**opnotpredefined %d", acc_op);
```

Comment:
I am thinking an assertion should work here since if the assertion is triggered, we probably should fix the code rather than return an error code to user.

Reply:
Again, changing this seems out of scope for this work.